### ISSUE TITLE
monitoring does not work for me

### ISSUE THREAD

**Issue by Torsten85** [packruler/rewrite-body#14]
**Describe the bug**
I've tried to use the `monitoring` feature, but it deactivates all rewrites

Either the README is wrong or I am missing something obvious here

**Relevant Config**
This config works:

```yaml
http:
  middlewares:
    rewrite-test:
      plugin:
        rewrite-body:
          rewrites:
            - regex: "foo"
              replacement: "bar"
```

this config simply does nothing and logs _Ignoring unsupported request_
even though its the default config according to the README:

```yaml
http:
  middlewares:
    rewrite-test:
      plugin:
        rewrite-body:
          rewrites:
            - regex: "foo"
              replacement: "bar"
          monitoring:
            methods:
              - GET
            types: 
              - text/html
```

**Expected behavior**
The monitoring should work

**Actual behavior**
it logs _Ignoring unsupported request_ for requests with content-type `text/html; charset=utf-8`

**Comment by packruler** [comment]
Interesting 🤔 thanks for reporting this! I will look into it soon!

**Comment by packruler** [comment]
I am having trouble reproducing this. Can you provide more context. I am particularly interested in the `Accept-Encoding` header in the request and the `Content-Encoding` header in the response.

**Comment by Torsten85** [comment]
Hello,
this is the request (coming from chrome):
```
GET / HTTP/1.1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Accept-Encoding: gzip, deflate
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8,de;q=0.7
Cache-Control: no-cache
Connection: keep-alive
Host: xxx
Pragma: no-cache
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36
```

and this is the response (without having *rewrite-body* involved):
```
HTTP/1.1 200 OK
Date: Mon, 12 Sep 2022 08:04:13 GMT
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Server: Apache
X-Content-Type-Options: nosniff
Expires: Sun, 19 Nov 1978 05:00:00 GMT
Cache-Control: no-cache, must-revalidate
X-Content-Type-Options: nosniff
Content-Language: de
X-Frame-Options: SAMEORIGIN
Permissions-Policy: interest-cohort=()
```

its weird that the requests works when no monitoring is configured but does not work with the default monitoring configuration

**Comment by packruler** [comment]
Thanks for more context! This is very weird 🤔

**Comment by packruler** [comment]
I am still struggling to recreate this issue locally for me. Can you provide more log details? Specifically the Traefik logs saying the request was ignored.

**Comment by mdklapwijk** [comment]
With some custom debug info; monitoring undefined and EnsureDefaults configTypes modified to 2 strings:
```
2022-11-18T16:33:36.986608+01:00 lbs-01 traefik[2825581]: Rewrite-Body | DEBUG2022/11/18 16:33:36 restricted.go:50: MDK-DBG-TYPES: [[text/css text/html]]
2022-11-18T16:33:36.986721+01:00 lbs-01 traefik[2825581]: Rewrite-Body | DEBUG2022/11/18 16:33:36 restricted.go:50: MDK-DBG-TYPE: [text/css]
2022-11-18T16:33:36.986785+01:00 lbs-01 traefik[2825581]: Rewrite-Body | DEBUG2022/11/18 16:33:36 restricted.go:50: MDK-DBG-TYPE: [text/html]
2022-11-18T16:33:36.986817+01:00 lbs-01 traefik[2825581]: Rewrite-Body | DEBUG2022/11/18 16:33:36 restricted.go:50: MDK-DBG-METHODS: [[GET]]
2022-11-18T16:33:36.986847+01:00 lbs-01 traefik[2825581]: Rewrite-Body | DEBUG2022/11/18 16:33:36 restricted.go:50: MDK-DBG-METHOD: [GET]
```

monitoring defined:
```
2022-11-18T16:30:01.944947+01:00 lbs-01 traefik[2824293]: Rewrite-Body | DEBUG2022/11/18 16:30:01 restricted.go:50: MDK-DBG-TYPES: [[║24║text/css║text/html║text/javascript]]
2022-11-18T16:30:01.945097+01:00 lbs-01 traefik[2824293]: Rewrite-Body | DEBUG2022/11/18 16:30:01 restricted.go:50: MDK-DBG-TYPE: [║24║text/css║text/html║text/javascript]
```

modifications to request_wrapper.go:
```
// SupportsProcessing determine if http.Request is supported by this plugin.
func (req *RequestWrapper) SupportsProcessing() bool {
        acceptHeader := req.Header.Get("Accept")
        isSupported := false

        req.logWriter.LogDebugf("MDK-DBG-TYPES: [%v]", req.monitoring.Types)

        for _, monitoredType := range req.monitoring.Types {
                req.logWriter.LogDebugf("MDK-DBG-TYPE: [%s]", monitoredType)
                if strings.Contains(acceptHeader, monitoredType) {
                        isSupported = true
                }
        }

        if !isSupported {
                return false
        }

        isSupported = false

        req.logWriter.LogDebugf("MDK-DBG-METHODS: [%v]", req.monitoring.Methods)

        // Ignore non GET requests
        for _, monitoredMethod := range req.monitoring.Methods {
                req.logWriter.LogDebugf("MDK-DBG-METHOD: [%s]", monitoredMethod)
                if strings.Contains(req.Method, monitoredMethod) {
                        isSupported = true
                }
        }

        if !isSupported {
                return false
        }

        if strings.Contains(req.Header.Get("Upgrade"), "websocket") {
                return false
        }

        return true
}
```

**Comment by mdklapwijk** [comment]
This fixes/worksaround it for me; httputil/monitoring_config.go:
```
func (config *MonitoringConfig) EnsureDefaults() {
        if len(config.Methods) == 0 {
                config.Methods = []string{http.MethodGet}
        }

        if len(config.Types) == 0 {
                config.Types = []string{"text/html"}
        }

        // handle mangled/flattened monitoring config
        if len(config.Methods) == 1 {
                for _, configMethod := range config.Methods {
                        if strings.Contains(configMethod,"║") {
                                config.Methods = strings.Split(strings.ReplaceAll(configMethod,"║24║",""),"║")
                        }
                }
        }

        if len(config.Types) == 1 {
                for _, configType := range config.Types {
                        if strings.Contains(configType,"║") {
                                config.Types = strings.Split(strings.ReplaceAll(configType,"║24║",""),"║")
                        }
                }
        }
}
```

As I lack any experience with golang, it is still completely unclear to me how the `[]string` gets mangled in the first place...

**Comment by packruler** [comment]
@mdklapwijk can you provide your Traefik config where you setup the extension? I am not sure what would cause that `║24║` string.

**Comment by mdklapwijk** [comment]
traefik:
```
root@lbs-01:/etc/traefik2# cat /etc/systemd/system/traefik2.service
[Unit]
Description=traefik2
Documentation=https://docs.traefik.io/
Wants=basic.target
After=basic.target network.target

[Service]
Type=simple
WorkingDirectory=/var/lib/traefik2
ExecStart=/usr/local/bin/traefik --configFile=/etc/traefik2/config.yaml
Restart=on-failure
RestartSec=30s

[Install]
WantedBy=multi-user.target
```
```
root@lbs-01:/etc/traefik2# l /usr/local/bin/traefik
lrwxrwxrwx 1 root root 39 Nov  8 16:24 /usr/local/bin/traefik -> /var/lib/traefik2/traefik-2.9.4/traefik
```
```
root@lbs-01:/etc/traefik2# uname -a
Linux lbs-01 5.10.0-19-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64 GNU/Linux
```
```
root@lbs-01:/etc/traefik2# cat /etc/debian_version
11.5
```

static:
```
root@lbs-01:/etc/traefik2# cat config.yaml
---
accesslog:
  filepath: "/var/log/traefik2/access.log"
  bufferingSize: 0
  fields:
    defaultmode: keep
    names:
      StartUTC: drop
    headers:
      defaultmode: keep
      names:
        User-Agent: keep
api:
  dashboard: true
entrypoints:
  http:
    address: ":80"
    http:
      redirections:
        entrypoint:
          permanent: true
          scheme: https
          to: https
  https:
    address: ":443"
    proxyprotocol:
      insecure: true
    forwardedheaders:
      insecure: true
  traefik:
    address: ":8443"
experimental:
  localplugins:
    rewrite-body:
      modulename: github.com/packruler/rewrite-body
log:
  filepath: "/var/log/traefik2/traefik.log"
  level: debug
providers:
  file:
    directory: "/etc/traefik2/dynamic.d"
    watch: true
serverstransport:
  insecureskipverify: true
```

dynamic:
```
root@lbs-01:/etc/traefik2/dynamic.d# cat tls.yaml
---
tls:
  certificates:
  - certfile: "/etc/ssl/certs/all.example.com.fullchain.crt"
    keyfile: "/etc/ssl/private/all.example.com.key"
  - certfile: "/etc/ssl/certs/all.int.example.com.fullchain.crt"
    keyfile: "/etc/ssl/private/all.int.example.com.key"
  options:
    default:
      minversion: VersionTLS12
      snistrict: true
    maxtls12:
      maxversion: VersionTLS12
      snistrict: true
    mintls11:
      minversion: VersionTLS11
      snistrict: true
    mintls12:
      minversion: VersionTLS12
      snistrict: true
    mintls13:
      minversion: VersionTLS13
      snistrict: true
  stores:
    default:
      defaultcertificate:
        certfile: "/etc/ssl/certs/all.example.com.fullchain.crt"
        keyfile: "/etc/ssl/private/all.example.com.key"
```
```
root@lbs-01:/etc/traefik2/dynamic.d# cat middlewares.yaml
---
http:
  middlewares:
    headers-default:
      headers:
        customrequestheaders:
          X-ID-LBS: KC-E23461C62429AG3229C79841B3AD3E3E-01
        customresponseheaders:
          X-ID-LBS: KC-E23461C62429AG3229C79841B3AD3E3E-01
```
```
root@lbs-01:/etc/traefik2/dynamic.d# cat router-api-09.example.com.yaml
---
http:
  routers:
    api-09.example.com:
      entrypoints:
      - https
      middlewares:
      - headers-default
      - headers-router-api-09.example.com
      - plugins-api-09.example.com
      - ratelimit-api-09.example.com
      priority: 4096
      rule: Host(`api-09.example.com`)
      service: api-09.example.com
      tls:
        options: mintls12
  middlewares:
    headers-router-api-09.example.com:
      headers:
        customrequestheaders:
          X-ID-ROUTER: KC-02AA8E954ADG7A134E51A9AE594DCC81-L7
          X-ID-ROUTER-GROUP: KC-CC46AE5428G34AFECFE8A9B9E7AC6928
        customresponseheaders:
          X-ID-ROUTER: KC-02AA8E954ADG7A134E51A9AE594DCC81-L7
          X-ID-ROUTER-GROUP: KC-CC46AE5428G34AFECFE8A9B9E7AC6928
    ratelimit-api-09.example.com:
      ratelimit:
        average: 100
        burst: 1000
        period: 60s
        sourcecriterion:
          ipstrategy:
            depth: 1
    plugins-api-09.example.com:
      plugin:
        rewrite-body:
          rewrites:
          - regex: api.example.com
            replacement: api-09.example.com
          - regex: media.example.com
            replacement: media-09.example.com
          monitoring:
            methods:
            - GET
            types:
            - application/javascript
            - application/json
            - text/css
            - text/html
            - text/javascript
            - text/json
  services:
    api-09.example.com:
      loadbalancer:
        healthcheck:
          followredirects: false
          hostname: api-09.example.com
          interval: 15s
          path: "/"
          scheme: https
          timeout: 3s
        servers:
        - url: https://192.168.1.151/
        - url: https://192.168.1.152/
        - url: https://192.168.1.153/
        - url: https://192.168.1.154/
        serverstransport: api-09.example.com
        sticky:
          cookie:
            domain: example.com
            httponly: true
            name: ROUTERID.CC46AE5428G34AFECFE8A9B9E7AC6928
            samesite: none
            secure: true
  serverstransports:
    api-09.example.com:
      servername: api-09.example.com
      insecureskipverify: false
      maxidleconnsperhost: 200
      forwardingtimeouts:
        dialtimeout: 15
        idleconntimeout: 15
        pingtimeout: 10
        readidletimeout: 15
        responseheadertimeout: 0
```

**Comment by packruler** [comment]
Thank you for that! I am not seeing anything that would make me expect that weird parsing but I'll try and replicate it soon

**Comment by packruler** [comment]
This should be resolved in [v1.2.0](https://github.com/packruler/rewrite-body/releases/tag/v1.2.0)
